server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java

```diff
@@ -395,6 +395,9 @@ public PipelinedRequest resolvePipeline(SearchRequest searchRequest, IndexNameEx
             if (searchRequest.pipeline() != null) {
                 // Named pipeline specified for the request
                 pipelineId = searchRequest.pipeline();
+            } else if (searchRequest.source() != null && searchRequest.source().pipeline() != null) {
+                // Inline pipeline specified for the request
+                pipelineId = searchRequest.source().pipeline();
```

Comment:
Rather than checking both properties, I think I would suggest making sure the `SearchRequest` and `SearchSourceBuilder` are in agreement. If you check `org.opensearch.rest.action.search.RestSearchAction#parseSearchSource`, you can see how URL parameters get pushed down into the `SearchSourceBuilder`. I would apply something similar here.

Alternatively, you could "lift" the value up into `SearchRequest` for backward compatibility. In `org.opensearch.rest.action.search.RestSearchAction#parseSearchRequest`, after the call to `parseSearchSource`, I think you could change [this line](https://github.com/opensearch-project/OpenSearch/blob/ab7e914c3956ac206cf269ca345574882770a283/server/src/main/java/org/opensearch/rest/action/search/RestSearchAction.java#L213) to:
```
searchRequest.pipeline(request.param("search_pipeline", searchRequest.source().pipeline()));
```

I think I like that second option more.

Reply:
Of course, unit testing it is mildly annoying. We don't have any unit tests for `RestSearchAction`. 

Adding test coverage for `parseSearchRequest` would be helpful, though.

Reply:
@msfroh wouldn't the rest request call of RestMultiSearchAction would go through the transport layer of SearchRequest [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/search/TransportSearchAction.java#L320)? This will not allow to read the params on the rest side of SearchAction. Wdyt?

Reply:
Oh -- you're right. The `SearchRequest` objects for `msearch` are constructed in [here](https://github.com/opensearch-project/OpenSearch/blob/6823ca1291ddcfc63d0704f93990560af1731e3f/server/src/main/java/org/opensearch/action/search/MultiSearchRequest.java#L224).

In theory, you could use the above line to lift the value from source into `SearchRequest` for `SearchAction` and `SearchTemplateAction`. Somewhere around [this line](https://github.com/opensearch-project/OpenSearch/blob/6823ca1291ddcfc63d0704f93990560af1731e3f/server/src/main/java/org/opensearch/action/search/MultiSearchRequest.java#L313) you could lift the value into `SearchRequest` from the `SearchSourceBuilder` as well.

Reply:
This looks like a better place. I have pulled the change in `MultiSearchRequest.java`